x-pack/filebeat/input/httpjson/request.go

```diff
@@ -146,38 +168,169 @@ func (rf *requestFactory) newHTTPRequest(stdCtx context.Context, trCtx *transfor
+	)
+	for i, rf := range r.requestFactories {
+		// iterate over collected ids from last response
+		if i == 0 {
+			// perform and store regular call responses
+			httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+			if err != nil {
+				return fmt.Errorf("failed to execute rf.collectResponse: %w", err)
+			}
+			if len(r.requestFactories) == 1 {
+				finalResps = append(finalResps, httpResp)
+				n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, finalResps, true)
+				if err != nil {
+					return err
+				}
+				continue
+			}
+			intermediateResps = append(intermediateResps, httpResp)
+			ids, err = r.getIdsFromResponses(intermediateResps, r.requestFactories[i+1].replace)
+			if err != nil {
+				return err
+			}
+			n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, intermediateResps, false)
+			if err != nil {
+				return err
+			}
+		} else {
+			if len(ids) == 0 {
+				n = 0
+				continue
+			}
+			urlCopy = rf.url
+			urlString = rf.url.String()
+			// perform request over collected ids
+			for _, id := range ids {
+				// reformat urls of requestFactory using ids
+				rf.url, err = generateNewUrl(rf.replace, urlString, id)
+				if err != nil {
+					return fmt.Errorf("failed to generate new URL: %w", err)
+				}
+
+				// collect data from new urls
+				httpResp, err = rf.collectResponse(stdCtx, trCtx, r, publisher)
+				if err != nil {
+					return fmt.Errorf("failed to execute rf.collectResponse: %w", err)
+				}
+				// store data according to response type
+				if i == len(r.requestFactories)-1 && len(ids) != 0 {
+					finalResps = append(finalResps, httpResp)
+				} else {
+					intermediateResps = append(intermediateResps, httpResp)
+				}
+			}
+			rf.url = urlCopy
+
+			var resps []*http.Response
+			if i < len(r.requestFactories) {
+				resps = finalResps
+			} else {
+				// The if comdition (i < len(r.requestFactories)) ensures this branch never runs to the last element
+				// of r.requestFactories, therefore r.requestFactories[i+1] will never be out of bounds.
+				ids, err = r.getIdsFromResponses(intermediateResps, r.requestFactories[i+1].replace)
+				if err != nil {
+					return err
+				}
+				resps = intermediateResps
+			}
+			split = r.responseProcessor.split
+			r.responseProcessor.split = rf.split
+			n, err = r.processAndPublishEvents(stdCtx, trCtx, publisher, resps, i < len(r.requestFactories))
+			if err != nil {
+				return err
+			}
+			r.responseProcessor.split = split
+		}
+	}
+
+	r.log.Infof("request finished: %d events published", n)
+
+	return nil
+}
+
+// getIdsFromResponses returns ids from responses
+func (r *requester) getIdsFromResponses(intermediateResps []*http.Response, replace *regexp.Regexp) ([]string, error) {
+	var b []byte
+	var ids []string
+	var err error
+	// collect ids from all responses
+	for _, resp := range intermediateResps {
+		resp := resp
+		if resp.Body != nil {
+			b, err = io.ReadAll(resp.Body)
+			resp.Body.Close()
```

Comment:
There are two tings worrying me here:
### 1. Read timeouts
Here we read the body of a request that was made on [L227](https://github.com/elastic/beats/blob/3aa2a1148592e44ca41b6c1dc81af9e0624182b6/x-pack/filebeat/input/httpjson/request.go#L227) or [L226](https://github.com/elastic/beats/blob/3aa2a1148592e44ca41b6c1dc81af9e0624182b6/x-pack/filebeat/input/httpjson/request.go#L264). The problem I see is the `for` loop on [L256](https://github.com/elastic/beats/blob/3aa2a1148592e44ca41b6c1dc81af9e0624182b6/x-pack/filebeat/input/httpjson/request.go#L256), if that loop takes too long to execute (the requests are slow or there are too many of them), for each request that was made Filebeat will keep the socket open and won't read any data until we reach this bit of the code. If the waiting time until we start reading those responses is longer than the set timeout (on Filebeat or on the remote server) then this read will fail.

### 2. The read and close of the body are quite far from where the request is made
Every time we make an HTTP request we need to make sure we close the response body, but because there are other parts of the code that could error (like [L258-L261](https://github.com/elastic/beats/blob/3aa2a1148592e44ca41b6c1dc81af9e0624182b6/x-pack/filebeat/input/httpjson/request.go#L258-L261)) before we reach the point where we read and close the response body, we cannot guarantee that will always happen. That can become a big problem because Filebeat will start leaking open file descriptors (sockets). Also the external system we're calling could suffer from a similar issue if they don't timeout/cancel the connection properly.

## A possible solution
To solve both issues, I suggest to refactor `collectResponse` so it reads (and closes) the response body right after the request is made, then it can return a `[]byte` instead of an `http.Response`. By what I could see in the code you don't use any other field of the response(is that right?). So I believe it should be safe to implement something like what I suggested.